### ISSUE TITLE
ngxAlias doesn't infer type correctly

### ISSUE THREAD

**Issue by snebjorn** [TNG/ngx-structurals#37]
Given that `isLoading$: Observable<boolean>`.
Then `as isLoading` in the below template is inferred as type `false | true | null`. But it can never be `null`.

```html
<ng-container *ngxAlias="(isLoading$ | async) === true as isLoading"></ng-container>
```

![image](https://user-images.githubusercontent.com/1266245/108492982-12c57480-72a6-11eb-854f-e4084b023923.png)

I'm reporting this here because `*ngIf` seems to work.

![image](https://user-images.githubusercontent.com/1266245/108493164-4c967b00-72a6-11eb-88e4-427e11d08f40.png)

Not sure if the problem is with `*ngxAlias` or the Angular Language Service

**Comment by Airblader** [comment]
Hi! Thanks for opening this issue. The union with `null` actually comes from the `async` pipe itself. The problem here is that your assumption

> But it can never be null.

is, in general, not correct. For `of(true)` that's the case, but for `of(true).pipe(delay(1000))` it isn't. The `ngIf` directive narrows the type because it doesn't render the template if the value is null, but ngxAlias exists for the very purpose of not conditionally rendering the template, so `null` is a valid type produced by `async`.

I'll close this as it both works as intended and is more of a topic of the `async` pipe, but feel free to ask if anything is unclear. If you search for "AsyncPipe null" in the issues of angular/angular, you'll find a bunch of issues revolving around this as well.

**Comment by snebjorn** [comment]
I'm comparing `boolean | undefined | null` from the AsyncPipe to `=== true` this assures the type is not `null`.

![image](https://user-images.githubusercontent.com/1266245/108503613-f8df5e00-72b4-11eb-943b-11046212eb18.png)

**Comment by Airblader** [comment]
Oh, whoops, I misread that. Let me reopen to look into this.

**Comment by Airblader** [comment]
Since you seem to be using VS Code, in the settings for the language service, do you have the experimental Ivy support enabled? If not, does it work if you enable it?

**Comment by snebjorn** [comment]
Sorry for not providing that info up front. Yes I'm using VSCode. I have the experimental Ivy support enabled.
Without that enabled `isLoading` is `any`

> Angular Language Service Version: 11.2.2
> 
> VSCode:
> Version: 1.53.2 (user setup)
> Electron: 11.2.1
> Chrome: 87.0.4280.141
> Node.js: 12.18.3
> V8: 8.7.220.31-electron.0
> OS: Windows_NT x64 10.0.19041

**Comment by Airblader** [comment]
This looks like an issue either with the language service or @angular/core to me. If you hover over the directive itself, the context type is correct:

```
declare (property) NgxAliasDirective<boolean>.ngxAlias: boolean
```

versus without `=== true`:

```
declare (property) NgxAliasDirective<boolean | null>.ngxAlias: boolean | null
```

The directive does already declare its context type to forward it correctly, which to my knowledge is really all I can do, see https://github.com/TNG/ngx-structurals/blob/master/projects/ngx-structurals/src/lib/ngx-alias/ngx-alias.directive.ts#L52-L55.

Would you mind opening an issue on angular/angular for this?

**Comment by snebjorn** [comment]
Not at all. I initially started typing this issue in the angular language service repo. But then I couldn't reproduce the issue using angular native directives. So figured they'd just send me your way :)

**Comment by Airblader** [comment]
I definitely know that the "built-in" directives used to have, and as far as I know still do have, some special treatment in the language service. Please do link the issue here as well after you opened it. Thanks!

**Comment by snebjorn** [comment]
https://github.com/angular/vscode-ng-language-service/issues/1137

**Comment by Airblader** [comment]
@all-contributors please add @snebjorn for bug

**Comment by Airblader** [comment]
@snebjorn Do you want to try if changing `NgxAliasContext#ngxAlias` to `ngxAlias: T = null!` (and likewise for `$implicit`) works as expected? Maybe also enable `strict: "true"` in the tsconfig. If so, you can PR it.

**Comment by snebjorn** [comment]
yay appears to work 🎉 

![image](https://user-images.githubusercontent.com/1266245/108779842-cdd06500-7567-11eb-9e37-6c7cc782ee9b.png)
![image](https://user-images.githubusercontent.com/1266245/108779861-d5900980-7567-11eb-8980-c72cd7cad2fa.png)

Not sure how to write a test for it though. As typescript doesn't really work in the html

**Comment by Airblader** [comment]
@all-contributors please add @snebjorn for code

**Comment by Airblader** [comment]
@snebjorn And this is now released in 1.2.1. Thanks for your contribution!

**Comment by snebjorn** [comment]
@Airblader thank you for the quick response and swift resolution :)
